testgrid/cmd/summarizer/summarizer.go

```diff
@@ -0,0 +1,173 @@
+	s := extractBasicSummary(queryData, dashboardName, dashboardTab, updateTimestamp)
+	s.FailingTestSummaries = extractTests(queryData)
+	return s, nil
+}
+
+func extractBasicSummary(queryData map[string]interface{}, dashboardName string, dashboardTab config.DashboardTab, updateTimestamp int64) summary.DashboardTabSummary {
+	var bugUrl string
+	bugTemplate, ok := queryData["open-bug-template"]
+	if !ok {
+		bugTemplate = ""
+	} else {
+		bugUrl = getStringInMap(bugTemplate.(map[string]interface{}), "url")
+	}
+
+	alerts := getStringInMap(queryData, "alerts")
+
+	lastRunTimestamp := 0
+	timestamp, ok := queryData["timestamps"]
+	if ok {
+		for i, e := range timestamp.([]int) {
+			if i == 0 || e > lastRunTimestamp {
+				lastRunTimestamp = e
+			}
+		}
+	}
+
+	latestGreen := getStringInMap(queryData, "latest-green")
+
+	overallStatus, ok := queryData["overall-status"]
+	if !ok {
+		overallStatus = summary.DashboardTabSummary_UNKNOWN
+	}
+
+	status := getStringInMap(queryData, "summary")
+
+	s := summary.DashboardTabSummary{
+		DashboardName:       dashboardName,
+		DashboardTabName:    dashboardTab.Name,
+		Alert:               alerts,
+		LastUpdateTimestamp: float64(updateTimestamp),
+		LastRunTimestamp:    float64(lastRunTimestamp),
+		LatestGreen:         latestGreen,
+		OverallStatus:       overallStatus.(summary.DashboardTabSummary_TabStatus),
+		Status:              status,
+		BugUrl:              bugUrl,
+	}
+
+	return s
+}
+
+func extractTests(queryData map[string]interface{}) []*summary.FailingTestSummary {
+	var failingTestSummaries []*summary.FailingTestSummary
+	tests, ok := queryData["tests"]
+	if !ok {
+		tests = []interface{}{}
+	} else {
+		for _, testItem := range tests.([]interface{}) {
+			test := testItem.(map[string]interface{})
+			testAlertMap, ok := test["alert"]
+			if ok {
+				testAlert := testAlertMap.(map[string]interface{})
+				target := getStringInMap(test, "target")
+				var linkedBugs []string
+				bugs, ok := test["linkedBugs"]
+				if !ok {
+					bugs = []int{}
+				}
+				for _, bugId := range bugs.([]int) {
+					linkedBugs = append(linkedBugs, string(bugId))
+				}
+
+				var failTestLink string
+				failTestId := getStringInMap(testAlert, "test-id")
+				if len(failTestId) > 0 {
+					var template map[string]interface{}
+					templateMap, ok := queryData["open-test-template"]
+					if !ok {
+						template = map[string]interface{}{}
+					} else {
+						template = templateMap.(map[string]interface{})
+						var options map[string]string
+						optionsMap, ok := template["options"]
+						if !ok {
+							options = map[string]string{}
+						} else {
+							options = optionsMap.(map[string]string)
+							var urlOptions string
+							if _, ok := options["target"]; ok {
+								options["target"] = target
+								urlOptions = fmt.Sprintf("target=%s", target)
+							}
+							if _, ok := options["id"]; ok {
```

Comment:
This is very indented. Could this be function be rearranged such that it isn't a deep if/else chain?

Reply:
I will sync up with Michelle on type first. I think we can make this cleaner and not so much nested if we can get static type on the input data. Thanks.